File: algorithms/medium/maximum_xor_for_each_query.ex
defmodule Solution do
  use Bitwise

  @spec get_maximum_xor(nums :: [integer], maximum_bit :: integer) :: [integer]
  def get_maximum_xor(nums, maximum_bit) do
    xor = Enum.reduce(nums, 0, fn n, acc -> bxor(acc, n) end)
    mask = (1 <<< maximum_bit) - 1

    nums
    |> Enum.reverse()
    |> Enum.reduce({[], xor}, fn n, {acc, xor} ->
      {[bxor(xor, mask) | acc], bxor(xor, n)}
    end)
    |> elem(0)
    |> Enum.reverse()
  end
end

File: algorithms/medium/maximum_xor_for_each_query.rs
impl Solution {
    pub fn get_maximum_xor(nums: Vec<i32>, maximum_bit: i32) -> Vec<i32> {
        let mut xor = nums.iter().fold(0, |acc, x| acc ^ x);
        let mut ans = Vec::new();
        let mask = (1 << maximum_bit) - 1;
        for i in 0..nums.len() {
            ans.push(xor ^ mask);
            xor ^= nums[nums.len() - i - 1];
        }
        ans
    }
}
